keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1055 @@
+
+
+### Risks and Mitigations
+
+##### Client Hanging Indefinitely
+
+With the `timeout` filter enabled today, it sends a `504 GatewayTimeout` response
+or an `error` to the client, if the request handler it is executing is blocking indefinitely.
+
+If the client does not enforce any timeout on its side, for example, not setting a
+global `Client` timeout, or not setting a deadline to the request `Context`; the client
+can hang forever waiting for a response from the server. This is certainly true for
+HTTP/1x where [net/http does not execute the request handler in a new goroutine](https://github.com/golang/go/blob/b8ac61e6e64c92f23d8cf868a92a70d13e20a124/src/net/http/server.go#L2031-L2039).
+
+> It's worth noting that by default `client-go` does not enforce any timeout globally.
+
+With the removal of the `timeout` filter, we need to ensure that a request handler that
+blocks on the server should never block a client that does not enforce timeout on its side.
+
+In fact, we can reproduce a "forever" hanging client with http/1x:
+```
+func TestClientHangingForeverWithHTTP1(t *testing.T) {
+	clientDoneCh, handlerDoneCh := make(chan struct{}), make(chan error, 1)
+	server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		defer close(handlerDoneCh)
+
+		ctrl := http.NewResponseController(w)
+		if err := ctrl.SetWriteDeadline(time.Now().Add(200 * time.Millisecond)); err != nil {
+			t.Errorf("expected no error from SetWriteDeadline, but got: %v", err)
+			return
+		}
+		if err := ctrl.SetReadDeadline(time.Now().Add(100 * time.Millisecond)); err != nil {
+			t.Errorf("expected no error from SetReadDeadline, but got: %v", err)
+			return
+		}
+
+		<-clientDoneCh
+	}))
+
+
+	defer server.Close()
+	server.StartTLS()
+
+	client := server.Client()
+	func() {
+		defer close(clientDoneCh)
+		client.Get(server.URL + "/foo")
+	}()
+
+	select {
+	case <-handlerDoneCh:
+	case <-time.After(wait.ForeverTestTimeout):
+		t.Errorf("expected the request handler to have terminated")
+	}
+}
+```
+Even though a read/write deadline is set, the client never receives any 
+response or error from the server. We have opened an [issue](https://github.com/golang/go/issues/65526)
+with the golang team to track this.
+
+Mitigation Steps:
+- a) If this is a bug, we can wait until the fix is in the net/http package.
+- b) Otherwise, we can enable this feature only for http/2.0 request(s), all
+  http/1x request(s) will continue to be served using the `Timeout` filter.
+
+We also need to add unit tests to verify that a blocked request handler
+does not block the corresponding client consequently:
+```
+neverDoneCh := make(chan struct{})
+func(w http.ResponseWriter, req *http.Request) {
+		<-neverDoneCh
+	}))
+```
+
+
+##### Request Handler Blocking Indefinitely:
+
+Once read/write deadline exceeds, the corresponding request handler goroutine will
+finish its run by returning an error, or throwing a panic as soon as it uses the
```

Comment:
I'm not clear here, will our code panic on the writer?  If so, we'll end up with stacks at the panic handler instead of a dedicated timeout handler?

If the code doesn't panic at the writer, how will we need to adjust our consuming code?

Reply:
> I'm not clear here, will our code panic on the writer?

That's an option, I describe it in the `Termination of the Request Handler` section. Basically, something like:
```
type timeoutWriter struct {
	http.ResponseWriter
}

func (tw *timeoutWriter) Write(p []byte) (int, error) {
	n, err := tw.ResponseWriter.Write(p)
	if errors.IsTimeout(err) {
		panic(http.ErrAbortHandler)
	}
	return n, err
}
```

`http.ErrAbortHandler`  will suppress the stack trace. 

The current timeout handler returns an ` http.ErrHandlerTimeout` error 
```
func (tw *baseTimeoutWriter) Write(p []byte) (int, error) {
	tw.mu.Lock()
	defer tw.mu.Unlock()

	if tw.timedOut {
		return 0, http.ErrHandlerTimeout
	}
        ....
}


>  If the code doesn't panic at the writer, how will we need to adjust our consuming code?

No, and this is in par with the timeout filter today, our code should return once it invokes `Write` and get an error, but if our code gets blocked or hung in the call chain, panic ensures a more reliable return path, thoughts?